File: software/python/visualizations/bci/systems/configuration/system_configuration.cpp
// Copyright 2004-present Facebook. All Rights Reserved.
// BCI C++ system configuration

#include "system_configuration.h"

#include <boost/math/constants/constants.hpp>
#include <logging/Checks.h>
#include <cmath>
#include <tuple>

namespace bci {
namespace systems {
namespace configuration {

SourceGroup::SourceGroup(const YAML::Node& sourcesNode, size_t sourcesOffset, double wavelength)
    : sourcesOffset_(sourcesOffset), numSources_(0), wavelength_(wavelength) {
  for (YAML::const_iterator sourceIterator = sourcesNode.begin();
       sourceIterator != sourcesNode.end();
       ++sourceIterator) {
    sources_.push_back(sourceIterator->as<YAML::Node>()["coordinate"].as<size_t>());
  }
  numSources_ = sources_.size();
}

Canister::Canister(
    const YAML::Node& canisterNode,
    size_t coordinatesOffset,
    size_t sourcesOffset,
    size_t detectorsOffset,
    const std::string& configVersion)
    : canisterName_(canisterNode["Name"].as<std::string>()),
      coordinatesOffset_(coordinatesOffset),
      sourcesOffset_(sourcesOffset),
      detectorsOffset_(detectorsOffset),
      numCoordinates_(0),
      numSources_(0),
      numDetectors_(0) {
  const YAML::Node& sourcesNode = canisterNode["Sources"];

  if (configVersion != "1.0.0") {
    XR_CHECK(sourcesNode["Wavelengths"], "Missing required field: Wavelengths");
  }
  if (sourcesNode["Wavelengths"]) {
    const YAML::Node& wavelengthsNode = sourcesNode["Wavelengths"];
    for (YAML::const_iterator wavelengthsIt = wavelengthsNode.begin();
         wavelengthsIt != wavelengthsNode.end();
         ++wavelengthsIt) {
      wavelengths_.emplace(
          wavelengthsIt->first.as<std::string>(), wavelengthsIt->second.as<double>());
    }
  }

  for (YAML::const_iterator sourceGroupIterator = sourcesNode.begin();
       sourceGroupIterator != sourcesNode.end();
       ++sourceGroupIterator) {
    // Skip wavelengths.
    if (sourceGroupIterator->first.as<std::string>() == "Wavelengths") {
      continue;
    }

    double wavelength = -1.0;
    // TODO(rsevile): Replace config string with 3 integers.
    if (configVersion != "1.0.0") {
      XR_CHECK(
          wavelengths_.count(sourceGroupIterator->first.as<std::string>()) == 1,
          "Missing required field {} in dict Wavelengths",
          sourceGroupIterator->first.as<std::string>());
    }
    if (wavelengths_.count(sourceGroupIterator->first.as<std::string>()) == 1) {
      wavelength = wavelengths_.at(sourceGroupIterator->first.as<std::string>());
    }
    // The numSources_ counter contains the current offset,
    // so it can be reused as the offset here.
    sourceGroups_.emplace(
        std::piecewise_construct,
        std::forward_as_tuple(sourceGroupIterator->first.as<std::string>()),
        std::forward_as_tuple(
            sourcesNode[sourceGroupIterator->first.as<std::string>()],
            sourcesOffset_ + numSources_,
            wavelength));

    numSources_ +=
        sourceGroups_.find(sourceGroupIterator->first.as<std::string>())->second.getNumSources();
  }
  coordinates_ = canisterNode["Coordinates"].as<std::vector<std::array<double, 3>>>();
  if (canisterNode["Transformation"]) {
    if (canisterNode["Transformation"]["AffineMatrix"]) {
      for (std::array<double, 3>& coordinate : coordinates_) {
        // Homogeneous coordinates affine transformation
        // http://graphics.cs.cmu.edu/nsp/course/15-462/Spring04/slides/04-transform.pdf
        matrix_multiply(
            coordinate,
            canisterNode["Transformation"]["AffineMatrix"]
                .as<std::array<std::array<double, 4>, 4>>());
      }
    } else if (canisterNode["Transformation"]["EulerAngles"]) {
      // Apply Euler angles transformation
      // https://en.wikipedia.org/wiki/Euler_angles
      // Rotations are applied in ZYX order first,
      // then translation is applied.

      // Apply rotation matrices
      // https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations
      std::array<double, 3> rotation =
          canisterNode["Transformation"]["EulerAngles"]["Rotation"].as<std::array<double, 3>>();

      if (canisterNode["Transformation"]["EulerAngles"]["Units"].as<std::string>() == "degrees") {
        for (int i = 0; i < 3; ++i) {
          rotation[i] = rotation[i] * boost::math::constants::pi<double>() / 180.0;
        }
      }

      std::array<double, 3> cosines = {
          std::cos(rotation[0]), std::cos(rotation[1]), std::cos(rotation[2])};

      std::array<double, 3> sines = {
          std::sin(rotation[0]), std::sin(rotation[1]), std::sin(rotation[2])};

      std::array<std::array<double, 3>, 3> zMatrix = {{
          {cosines[2], -sines[2], 0.0},
          {sines[2], cosines[2], 0.0},
          {0.0, 0.0, 1.0},
      }};

      std::array<std::array<double, 3>, 3> yMatrix = {{
          {cosines[1], 0.0, -sines[1]},
          {0.0, 1.0, 0.0},
          {-sines[1], 0.0, cosines[1]},
      }};

      std::array<std::array<double, 3>, 3> xMatrix = {{
          {1.0, 0.0, 0.0},
          {0.0, cosines[0], -sines[0]},
          {0.0, sines[0], cosines[0]},
      }};

      std::array<std::array<double, 3>, 3> yzMatrix;

      matrix_multiply(yzMatrix, yMatrix, zMatrix);

      std::array<std::array<double, 3>, 3> staticMatrix;

      matrix_multiply(staticMatrix, xMatrix, yzMatrix);

      std::array<double, 3> translation =
          canisterNode["Transformation"]["EulerAngles"]["Translation"].as<std::array<double, 3>>();

      for (std::array<double, 3>& coordinate : coordinates_) {
        matrix_multiply(coordinate, staticMatrix);
        for (int i = 0; i < 3; ++i) {
          coordinate[i] += translation[i];
        }
      }
    }
  }
  for (YAML::const_iterator detectorIterator = canisterNode["Detectors"].begin();
       detectorIterator != canisterNode["Detectors"].end();
       ++detectorIterator) {
    detectors_.push_back(detectorIterator->as<YAML::Node>()["coordinate"].as<size_t>());
  }
  numCoordinates_ = coordinates_.size();
  numDetectors_ = detectors_.size();
}

void Canister::matrix_multiply(
    std::array<std::array<double, 3>, 3>& result,
    const std::array<std::array<double, 3>, 3>& leftMatrix,
    const std::array<std::array<double, 3>, 3>& rightMatrix) {
  for (int row = 0; row < 3; ++row) {
    // Dot product with the row
    for (int column = 0; column < 3; ++column) {
      result[row][column] = 0.0;
      for (int i = 0; i < 3; ++i) {
        result[row][column] += leftMatrix[row][i] * rightMatrix[i][column];
      }
    }
  }
}

void Canister::matrix_multiply(
    std::array<double, 3>& coordinate,
    const std::array<std::array<double, 3>, 3>& matrix) {
  std::array<double, 3> result_coordinate = {0.0, 0.0, 0.0};

  for (int row = 0; row < 3; ++row) {
    // Dot product with the row
    for (int column = 0; column < 3; ++column) {
      result_coordinate[row] += matrix[row][column] * coordinate[column];
    }
  }

  for (int i = 0; i < 3; ++i) {
    coordinate[i] = result_coordinate[i];
  }
}

void Canister::matrix_multiply(
    std::array<double, 3>& coordinate,
    const std::array<std::array<double, 4>, 4>& matrix) {
  std::array<double, 4> homogeneous_coordinate = {0.0, 0.0, 0.0, 0.0};

  for (int row = 0; row < 4; ++row) {
    // Dot product with the row
    for (int column = 0; column < 4; ++column) {
      if (column < 3) {
        homogeneous_coordinate[row] += matrix[row][column] * coordinate[column];
      } else {
        homogeneous_coordinate[row] += matrix[row][column];
      }
    }
  }

  for (int i = 0; i < 3; ++i) {
    coordinate[i] = homogeneous_coordinate[i] / homogeneous_coordinate[3];
  }
}

SystemConfiguration::SystemConfiguration(const YAML::Node& rootNode)
    : configVersion_(rootNode["ConfigVersion"].as<std::string>()),
      deviceName_(rootNode["DeviceName"].as<std::string>()),
      numCoordinates_(0),
      numSources_(0),
      numDetectors_(0) {
  for (const auto& canisterNode : rootNode["Canisters"]) {
    // The counters contain the current offsets, so they
    // can be reused as offset counters too.
    canisters_.emplace_back(
        canisterNode, numCoordinates_, numSources_, numDetectors_, configVersion_);

    numCoordinates_ += canisters_.back().getNumCoordinates();
    numSources_ += canisters_.back().getNumSources();
    numDetectors_ += canisters_.back().getNumDetectors();
  }
}

} // namespace configuration
} // namespace systems
} // namespace bci

File: software/python/visualizations/bci/systems/configuration/system_configuration.py
#!/usr/bin/env python3
# Copyright 2004-present Facebook. All Rights Reserved.

import math
from dataclasses import InitVar, dataclass, field
from typing import Any, Dict, List

import numpy as np


@dataclass
class SourceGroup(object):
    """
    A group of sources of the same type.
    The meaning of "type" is dependent on the system.
    For DOT systems "type" indicates wavelength, e.g. "Red" or "Infrared".

    sources: A list of source coordinate indices.
    source_offset: The offset into a global sources array
    num_sources: The total number of sources.
    wavelength: Frequency of the light wave.
    """

    sources: np.ndarray
    source_offset: int
    num_sources: int = field(init=False)
    wavelength: float

    def __post_init__(self) -> None:
        self.num_sources = np.shape(self.sources)[0]

    def __getitem__(self, key: int) -> Any:
        return self._sources[key]

    @property
    def slice(self) -> Dict[str, slice]:
        return {
            "source": slice(self.source_offset, self.source_offset + self.num_sources)
        }


@dataclass
class Canister(object):
    """
    A canister.

    name: The name of the canister.
    wavelengths: A dictionary of wavelength name to wavelength frequency.
    coordinates: An array of 3D coordinates with shape (#sources, 3).
    source_groups: A list of source groups.
    detectors: An index array which represents the detector to coordinate mapping.
    coordinates_offset: The offset into a global coordinates array
    source_offset: The offset into a global sources array
    detector_offset: The offset into a global detectors array
    num_coordinates: The total number of coordinates.
    num_sources: The total number of sources.
    num_detectors: The total number of detectors.
    """

    name: str
    wavelengths: Dict[str, float]
    coordinates: np.ndarray
    source_groups: Dict[str, SourceGroup]
    detectors: np.ndarray
    coordinates_offset: int
    source_offset: int
    detector_offset: int
    num_coordinates: int = field(init=False)
    num_sources: int = field(init=False)
    num_detectors: int = field(init=False)

    def __post_init__(self) -> None:
        self.num_coordinates = np.shape(self.coordinates)[0]

        num_sources = 0
        for _, source_group in self.source_groups.items():
            num_sources += source_group.num_sources
        self.num_sources = num_sources

        self.num_detectors = np.shape(self.detectors)[0]

    @property
    def slice(self) -> Dict[str, slice]:
        return {
            "source": slice(self.source_offset, self.source_offset + self.num_sources),
            "detector": slice(
                self.detector_offset, self.detector_offset + self.num_detectors
            ),
        }


@dataclass
class SystemConfiguration(object):
    """
    A full system configuration.

    config_version: Config version.
    name: Device name.
    canisters: A list of canisters.
    num_coordinates: The total number of coordinates.
    num_sources: The total number of sources.
    num_detectors: The total number of detectors.
    """

    config_version: str = field(init=False)
    name: str = field(init=False)
    canisters: List[Canister] = field(init=False)
    num_coordinates: int = field(init=False)
    num_sources: int = field(init=False)
    num_detectors: int = field(init=False)
    num_measurements: int = field(init=False)
    _raw_configuration: InitVar[Dict[str, Any]]

    # Homogeneous coordinates affine transformation
    # This uses a left-multiplication by the matrix.
    # http://graphics.cs.cmu.edu/nsp/course/15-462/Spring04/slides/04-transform.pdf
    def homogeneous_matrix_transformation(
        self, matrix: np.ndarray, coordinate: np.ndarray
    ) -> np.ndarray:
        # Append w coordinate
        coordinate = np.append(coordinate, 1.0)
        coordinate = np.matmul(matrix, coordinate)

        # Normalize by w coordinate
        coordinate[0:3] /= coordinate[3]

        return coordinate[0:3]

    def apply_affine_matrix(
        self,
        coordinates: np.ndarray,
        canister_coordinates: Dict[str, Any],
        affine_matrix: Dict[str, Any],
    ):
        matrix = np.array(affine_matrix)
        for c, coordinate in enumerate(canister_coordinates):
            coordinate_array = np.array(coordinate)
            assert np.shape(matrix) == (4, 4)
            assert np.shape(coordinate_array) == (3,)
            coordinate = self.homogeneous_matrix_transformation(
                matrix, coordinate_array
            )
            coordinates[c] = coordinate

    def apply_euler_angles(
        self,
        coordinates: np.ndarray,
        canister_coordinates: Dict[str, Any],
        euler_angles: Dict[str, Any],
    ):
        # Apply Euler angles transformation
        # https://en.wikipedia.org/wiki/Euler_angles
        # Rotations are applied in ZYX order first,
        # then translation is applied.

        # Apply rotation matrices
        # https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations

        x = euler_angles["Rotation"][0]
        y = euler_angles["Rotation"][1]
        z = euler_angles["Rotation"][2]

        if euler_angles["Units"] == "degrees":
            x = math.radians(x)
            y = math.radians(y)
            z = math.radians(z)

        cos_z = math.cos(z)
        sin_z = math.sin(z)
        z_matrix = np.array(
            [[cos_z, -sin_z, 0.0], [sin_z, cos_z, 0.0], [0.0, 0.0, 1.0]]
        )

        cos_y = math.cos(y)
        sin_y = math.sin(y)
        y_matrix = np.array(
            [[cos_y, 0.0, -sin_y], [0.0, 1.0, 0.0], [-sin_y, 0.0, cos_y]]
        )

        cos_x = math.cos(x)
        sin_x = math.sin(x)
        x_matrix = np.array(
            [[1.0, 0.0, 0.0], [0.0, cos_x, -sin_x], [0.0, sin_x, cos_x]]
        )

        static_transformation = np.matmul(x_matrix, np.matmul(y_matrix, z_matrix))

        for c, coordinate in enumerate(canister_coordinates):
            coordinate_array = np.array(coordinate)

            coordinate_array = np.matmul(static_transformation, coordinate_array)

            for x in range(3):
                coordinate_array[x] += euler_angles["Translation"][x]

            coordinates[c] = coordinate_array

    def apply_transformation(
        self,
        coordinates: np.ndarray,
        canister_coordinates: Dict[str, Any],
        transformation: Dict[str, Any],
    ):
        if "AffineMatrix" in transformation:
            self.apply_affine_matrix(
                coordinates, canister_coordinates, transformation["AffineMatrix"]
            )
        elif "EulerAngles" in transformation:
            self.apply_euler_angles(
                coordinates, canister_coordinates, transformation["EulerAngles"]
            )

    def __post_init__(self, _raw_configuration: Dict[str, Any]) -> None:
        # Parse raw config to build the system configuration
        self.config_version = _raw_configuration["ConfigVersion"]
        self.name = _raw_configuration["DeviceName"]
        self.canisters = []

        coordinates_offset = 0
        source_offset = 0
        detector_offset = 0

        for canister in _raw_configuration["Canisters"]:
            coordinates = np.full(
                (len(canister["Coordinates"]), 3), np.nan, dtype=np.float64
            )
            # Apply transformation
            if "Transformation" in canister.keys():
                self.apply_transformation(
                    coordinates, canister["Coordinates"], canister["Transformation"]
                )
            else:
                for c, coordinate in enumerate(canister["Coordinates"]):
                    coordinates[c] = coordinate

            wavelengths = {}
            # TODO(rsevile): Replace config version with 3 integers.
            if self.config_version != "1.0.0":
                assert "Wavelengths" in canister["Sources"]

            if "Wavelengths" in canister["Sources"]:
                wavelengths = canister["Sources"]["Wavelengths"]

            total_sources = 0
            source_groups = {}
            for key, raw_sources in canister["Sources"].items():
                if key == "Wavelengths":
                    continue
                sources_array = np.full(len(raw_sources), 0, np.int32)
                for s, source in enumerate(raw_sources):
                    sources_array[s] = source["coordinate"]
                if self.config_version == "1.0.0":
                    source_groups[key] = SourceGroup(
                        sources_array,
                        source_offset + total_sources,
                        wavelengths[key] if key in wavelengths else -1.0,
                    )
                else:
                    source_groups[key] = SourceGroup(
                        sources_array, source_offset + total_sources, wavelengths[key]
                    )
                total_sources += len(raw_sources)

            detectors = np.full(len(canister["Detectors"]), 0, np.int32)
            for d, detector in enumerate(canister["Detectors"]):
                detectors[d] = detector["coordinate"]

            self.canisters.append(
                Canister(
                    name=canister["Name"],
                    wavelengths=wavelengths,
                    coordinates=coordinates,
                    source_groups=source_groups,
                    detectors=detectors,
                    coordinates_offset=coordinates_offset,
                    source_offset=source_offset,
                    detector_offset=detector_offset,
                )
            )
            coordinates_offset += np.shape(coordinates)[0]
            source_offset += total_sources
            detector_offset += np.shape(detectors)[0]

        num_coordinates = 0
        for canister in self.canisters:
            num_coordinates += canister.num_coordinates
        self.num_coordinates = num_coordinates

        num_sources = 0
        for canister in self.canisters:
            num_sources += canister.num_sources
        self.num_sources = num_sources

        num_detectors = 0
        for canister in self.canisters:
            num_detectors += canister.num_detectors
        self.num_detectors = num_detectors

        self.num_measurements = self.num_detectors * self.num_sources
